modules/node-shared/src/main/scala/org/tessellation/node/shared/domain/statechannel/StateChannelValidator.scala

```diff
@@ -34,20 +34,40 @@ trait StateChannelValidator[F[_]] {
   def validate(
     stateChannelOutput: StateChannelOutput,
     snapshotOrdinal: SnapshotOrdinal,
-    staked: Balance
+    snapshotFeesInfo: SnapshotFeesInfo
   )(implicit hasher: Hasher[F]): F[StateChannelValidationErrorOr[StateChannelOutput]]
+
   def validateHistorical(
     stateChannelOutput: StateChannelOutput,
     snapshotOrdinal: SnapshotOrdinal,
-    staked: Balance
+    snapshotFeesInfo: SnapshotFeesInfo
   )(
     implicit hasher: Hasher[F]
   ): F[StateChannelValidationErrorOr[StateChannelOutput]]
-
 }
 
 object StateChannelValidator {
 
+  def validateIfAddressAlreadyUsed(
+    metagraphId: Address,
+    existingFeesAddresses: Map[Address, Set[Address]],
+    maybeAddress: Option[Address]
+  ): StateChannelValidationErrorOr[Unit] = maybeAddress match {
+    case None => ().validNec
+    case Some(address) =>
+      val metagraphAddresses = existingFeesAddresses.getOrElse(metagraphId, List.empty)
```

Comment:
You can omit constructing empty list as a fallback by directly using `exists` on `Option`. Let's save some memory ;)

```scala
val existsInMetagraphAddresses = existingFeesAddresses.get(metagraphId).exists(_.contains(address))
        
if (existsInMetagraphAddresses) {
  stateChannelOutput.snapshotBinary.validNec
} else {
  ...
}
```

Reply:
Also one addition:

We have `Set[Address]` but you made a fallback to `List[Address]`.
These 2 structures are different:

- List has it's specific order and can have duplicates and `contains` is `O(n)`
- Set doesn't have any particular order and `contains` is `O(1)` or `O(log n)`

In general in Scala we should be careful with using proper structures :+1:

Also some interesting reference if you want to read more: https://docs.scala-lang.org/overviews/collections-2.13/performance-characteristics.html

Reply:
Cool, thanks =]